Advisor message history window

ShiroAka opened the pull request:
*Fixes #530 *

**Describe what the proposed change does**
- Adds a new dialog which lets the user see a history of the adviser's messages, which displays latest messages first
- The new dialog can be opened via the "Charts" menu, and also by using the hotkeys (by default CTRL + H)

<img width="525" height="351" alt="image" src="https://github.com/user-attachments/assets/1e50093c-1650-41a5-a454-d8eea675a88d" />

As of now, the dialog lets users delete single messages (by clicking the red button next to each message), or clear the whole history (by clicking the header above the red buttons).

ShiroAka committed:
Add dialog to see advisor message history

ShiroAka committed:
Add missing hotkey handler

ShiroAka committed:
Attempt to fix build errors

Matroftt commented:
Add a screenshot of this menu and tell how it could be opened in "describe what the proposed change does"


Also, are you sure I made this menu? :)

ShiroAka committed:
Fix copyright

Alberth289346 commented:
Align the top of the messages list with the top of the 'x' at the right?

With respect to messages:
- What happens if the same message appears again?
- Do messages auto-expire after some time?
- Would it be useful to know when a message appeared?

ShiroAka commented:
> Align the top of the messages list with the top of the 'x' at the right?
> 
> With respect to messages:
> 
> * What happens if the same message appears again?
> * Do messages auto-expire after some time?
> * Would it be useful to know when a message appeared?

As seen in ARGAMX's screenshot, messages repeat without limits. I just came up with the idea to bring a message to the top of the list if was already in it.
But for that I should implement a way to identify the messages (ideally without having to compare strings, which is already done in some other parts of the code).

Alternatively I could still let messages repeat, but avoid adding it to the history if a similar message was added X time ago.

There is no message expiration; I don't see the benefit of adding that.
There's only a max. amount of messages that can be added to the history and, if the limit is reached, the oldest messages are replaced.

About knowing when a message appeared, if implemented correctly, it may be useful. Perhaps I could add another row showing how many seconds (or minutes) ago was the message sent.
The thing is that in order to add that info I should make the table a bit wider (some didn't like this idea) or increase the row height (which would add a lot of free space for most messages and wouldn't look nice, IMO)

ARGAMX commented:
> About knowing when a message appeared, if implemented correctly, it may be useful. Perhaps I could add another row showing how many seconds (or minutes) ago was the message sent.
> The thing is that in order to add that info I should make the table a bit wider (some didn't like this idea) or increase the row height (which would add a lot of free space for most messages and wouldn't look nice, IMO)

The message history window is already quite functional even without the message posting time.
But if we add the time, then given that time in the game flows at different speeds depending on the selected speed, how relevant will that time be?
In my opinion, in this case, it makes sense to display only the in-game date.

<img width="587" height="176" alt="Screenshot 2025-09-18 at 09 20 35" src="https://github.com/user-attachments/assets/d9d05a4b-fcc6-4916-8b9d-4d7b2c838c90" />

lewri commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,181 @@
+local ticks_to_skip = 5
+local tick_timer = ticks_to_skip
+
+local row_height = 40
```
It might be possible to make your row heights dynamic with some help from the draw functions on text (I think UIInformation and UIConfirmation do this). But I need to look into this patch deeper as well.

ShiroAka commented:
> I wouldn't deduplicate, this is the real notification history.

But maybe I should filter certain messages that aren't very useful, like the errors when placing staff.

ShiroAka commented:
> The message history window is already quite functional even without the message posting time. But if we add the time, then given that time in the game flows at different speeds depending on the selected speed, how relevant will that time be? In my opinion, in this case, it makes sense to display only the in-game date.

In a ledger it makes sense to display the in-game date because it refers to what's actually happening in the simulation, but for the notifications I'm not so sure. 

The way I see it, the notifications are more related to general advice or results of your direct actions, and are something that could impact your actions directly (instead of a being a direct result of the simulation). So the focus is not on the simulation per-se, but on the user. That's why I suggested displaying how many real seconds/minutes ago did the notification appear, instead of tracking that based on the simulation's/in-game time.

But you have a good point on the game speed. I thought about that too.

ShiroAka committed:
Add adviser message filter

ARGAMX commented:
> In a ledger it makes sense to display the in-game date because it refers to what's actually happening in the simulation, but for the notifications I'm not so sure.
>
> The way I see it, the notifications are more related to general advice or results of your direct actions, and are something that could impact your actions directly (instead of a being a direct result of the simulation). So the focus is not on the simulation per-se, but on the user. That's why I suggested displaying how many real seconds/minutes ago did the notification appear, instead of tracking that based on the simulation's/in-game time.

Yes, sounds logical.

The thing is that as I think most user-initiated notifications make less sense to store in history.
Like those messages about trying to put an employee in an inapropriate tile.
Or when we put a nurse to the GP office and the Adviser says we can't put her here.
The nurse will leave the office on her own.

But why should we reread that type of messages in history when the moment has already passed?
For such messages, I agree, game time really doesn't matter.
But I don't think there's much point in showing these kinds of messages in history at all. 
They lose relevance instantly.

When we talk about Advisor notifications, I actually think primarily about events in the simulation.

Why them? Because the player isn't the cause of these notifications. They pop up when the player is busy with buildng the room and can't switch to another task or something else and misses them. That's where a notification history would be most useful. I constantly miss messages like these as a player.

And in light of this problem, your implementation of the history for such notifications is incredibly useful!

I thought you were trying to solve this exact problem.
This is exactly the problem I'm experiencing as a player.

What I'm constantly missing:

Student gained X skill or degree (result of the simulation)
New room researched (result of the simulation)
Some research direction is fully researched (result of the simulation)
Boiler breakdown (result of the simulation)
You've reached the balance plus and it's time to pay off the bank loan (result of the simulation)

That's why the question:
What specific general advices or results of your direct actions would you like to keep in history?

Alberth289346 commented:
> ... But why should we reread that type of messages in history when the moment has already passed?

I typically notice the advisor when he dives down behind the bottom window again, and completely miss the message.

New users may not immediately catch what is wrong, and currently you cannot get the message back for more careful reading.

Different users have different needs. By default log everything. Then add settings to skip logging some messages. In that way everybody can pick what they prefer.

Alberth289346 commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,181 @@
+  -- Calculate menu's height
+  local height_divisor = 4
+  local approx_rows_space = math.floor(app.config.height / height_divisor)
+  self.rows = math.ceil(approx_rows_space / row_height)
```
This is actually `self.num_visible_rows` ?

`rows` is a bit too generic. The name doesn't make it clear what is stored in it. (Eg the name does not exclude rows of messages).

Alberth289346 commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,181 @@
+  self.rows = math.ceil(approx_rows_space / row_height)
+  local rows_space = self.rows * row_height
+
+  local width = 500
```
You may want to move this line to the other constants at lines 36-38.
In that way, all magic sizes are near each other. No need to go look through the entire file to find them then.

ShiroAka commented:
I just had an idea regarding the message filtering (similar to what Alberth proposed above)

So far the staff placement messages are being filtered from the history, and I have been thinking about adding more messages to the filter...

But what if I add a column that lets the user block a certain message from the history?
Blocking a message should delete all other instances from the history and prevent it from appearing again.

The only issue is that I would need a way to identify the messages once they are in the history (not necessarily with an ID).

And I would also need to add a section somewhere in the config to unblock the messages (and possibly a way to block them from there)

ARGAMX reviewed (commented):
I merged this change for myself and played with it.
This is a very useful enchahment! 👍
I enjoyed playing with it!

What I didn't like was that after loading, I can't see the Adviser history.
Sometimes I want to look at something in another save and then go back and continue playing current level.
So I save, load another save, then load my first save — now the Adviser message history is empty — and then I regret that I loaded the game.
So, saving the game doesn't save this state, which isn't very good or convenient.

On the other hand, if enable displaying the history after loading, then when switching languages, you might get artifacts like the text "?????????".

But I think the damage from this defect is less than the damage from loosing the history.
Overall, I recommend keeping the history after loading.
Otherwise, this cool feature loses some of its usefulness.
I have disabled history wipe on load for me localy.

I hope this PR haven't hung on?

ShiroAka committed:
Replaced message filter and improved code

- Replaced the messages filter in the adviser's message history for a method that brings existing messages to the top of the list

- Improved the code related to this feature to make it easier to read/understand

ARGAMX reviewed (approved):
![_chuck-norris](https://github.com/user-attachments/assets/a05e4d0a-86a0-4e73-ac34-b9606cfb069b)

ShiroAka committed:
Improve update algorithm

ShiroAka committed:
Remove tick vars

ARGAMX committed:
added newline at the end of adviser_history.lua

ARGAMX commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,182 @@
+
+function UIAdviserHistory:close()
+  Window.close(self)
+end
```
> newline nit

newline added 👌

ShiroAka commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,191 @@
+
+function UIAdviserHistory:close()
+  Window.close(self)
+end
```
I added the newline but it seems my github client chose to remove it (or maybe it was VSCode) :S

ARGAMX replied:
> I added the newline but it seems my github client chose to remove it (or maybe it was VSCode) :S

Interesting. 🙂
If this is the case then I'm sure there must be an option in the settings to turn this on/off.

But specifically for this PR, if a new line has already been added, everything is ok here now.

lewri commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,192 @@
+  self.esc_closes = true
+  self.resizable = false
+
+  self.adviser_messages = {} -- Contains all the messages in the adviser.adviser_history prop
```
What's the prop for/mean?

ShiroAka replied:
Property/attribute
But I'll be removing that list, so this is not relevant anymore

lewri reviewed (commented):
Please luadoc the functions of your new file. I believe you're mostly there though.

ARGAMX committed:
Merge branch 'master' into feature/adviserMessageHistory

ARGAMX commented on CorsixTH/Lua/config_finder.lua:
```diff
@@ -828,7 +829,8 @@ param(hotkeys_values, 'ingame_panel_research') ..
 param(hotkeys_values, 'ingame_panel_policy') ..
-param(hotkeys_values, 'ingame_panel_machineMenu') ..
+param(hotkeys_values, 'ingame_panel_adviserHistory') ..
+param(hotkeys_values, 'ingame_panel_machineMenu') ..  
```
```suggestion
param(hotkeys_values, 'ingame_panel_machineMenu') ..
param(hotkeys_values, 'ingame_panel_adviserHistory') ..
```

ARGAMX committed:
master merge fix

ShiroAka commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,192 @@
+end
+
+function UIAdviserHistory:update()
+  -- self.adviser_messages = self.ui.adviser.message_history --> WRONG, this way both variables point to the same table
```
old comment that I made while refactoring and forgot to delete

ShiroAka commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,192 @@
+    local message = self.adviser_messages[i + scroll_pos - 1]
+    local row = self.list_table[i]
+    if message then
+      row.message_panel:setLabel(" " .. message, nil, "left"):setColour(col_bg)
```
I added that space to avoid the text being so close to the edge of the column. I feel it's easier to read with that extra space
Here you can clearly see the difference
<img width="1074" height="342" alt="image" src="https://github.com/user-attachments/assets/3032a49b-e605-4c65-a92c-5c78665e4280" />

lewri replied:
I think you've managed to somehow recreate #2467 here. Really, your spacing should be equal for each line.

ARGAMX replied:
I like the space ShiroAka added here.
I agree with him that it's easier to read with that extra space.
So I wouldn't want to remove it either.

TheCycoONE replied:
It's a hack and not found anywhere else in CorsixTH. Can keep it but add an issue to replace it with padding support for panel labels.

lewri replied:
I'd rather we kept a uniform method to update the paragraph formatting and not pick and choose as a hack. I'd recommend the alternative of removing this hack and adding a universal controller after (someone then has more reason to actually do the work).

ARGAMX replied:
Ok, that make sense. 
That space removed. 👌

ShiroAka committed:
Code fixes/refactors

ShiroAka committed:
Add spanish translations

ShiroAka commented on CorsixTH/Lua/languages/english.lua:
```diff
@@ -1034,6 +1036,21 @@ tooltip.machine_menu = {
   close = "Close the machine list dialog",
 }
 
+adviser_history = {
+  message = "Message",
+  close = "Close",
+}
+
+tooltip.adviser_history = {
+  delete_message = "Click to dismiss this message",
+  message = "List of adviser messages - Newest messages appear first",
+  header = {
+    message = "Adviser messages",
+    delete_message = "Click to delete all messages"
+  },
+  close = "Close the adviser history dialog",
+}
```
Now they are :)

ShiroAka commented on CorsixTH/Lua/dialogs/bottom_panel.lua:
```diff
@@ -881,6 +883,16 @@ function UIBottomPanel:dialogMachineMenu(enable)
   self.ui:playSound("selectx.wav")
 end
 
+function UIBottomPanel:dialogAdviserHistory(enable)
```
I can add a button later, but not in this PR

tobylane commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,187 @@
+  self.delete_col_width = 30
+  self.message_col_width = 400
+
+  local total_width = 500
```
```suggestion
  local total_width = 489
```
The margin + columns + line 132 spacing + scrollbar width + margin is 489, 500 looks too wide.

ARGAMX replied:
> The margin + columns + line 132 spacing + scrollbar width + margin is 489, 500 looks too wide.

I think so too. Applied

tobylane commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,187 @@
+        :makeScrollbar(col_bg, scrollbarMovedCallback, 1, 1, 10, 1)
+    -- Add close button
+    self:addPanel(337, x,  8):makeButton(0, 0, 24, 24, 338, self.close)
+        :setTooltip(_S.tooltip.adviser_history.close)
```
```suggestion
    y = self.window_margin + self.header_height + 5

    -- Add scrollbar
    self.scrollbar = self:addColourPanel(x, y, 24, self.row_height * rows - 5,
        col_shadow.red, col_shadow.green, col_shadow.blue)
        :makeScrollbar(col_bg, scrollbarMovedCallback, 1, 1, 10, 1)
    -- Add close button
    self:addPanel(337, x, 14):makeButton(0, 0, 24, 24, 338, self.close)
        :setTooltip(_S.tooltip.adviser_history.close)
```
Not sure on this suggestion. Is the close button better off with the middle or bottom matching the column header?

ARGAMX replied:
> Not sure on this suggestion. Is the close button better off with the middle or bottom matching the column header?

I think so too. Applied

tobylane commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,187 @@
+  end
+end
+
+function UIAdviserHistory:close()
```
Not necessary as it's inherited.

ARGAMX replied:
Applied

tobylane commented on CorsixTH/Lua/game_ui.lua:
```diff
@@ -1318,6 +1318,8 @@ function GameUI:afterLoad(old, new)
     self.subtitles = Subtitles(self)
     self:addWindow(self.subtitles)
+    self.adviser.message_history = {}
+    self.adviser.message_history_limit = 20
```
This can go in `UIAdviser:afterLoad`, with a savegame version bump in App.

ARGAMX replied:
> This can go in `UIAdviser:afterLoad`, with a savegame version bump in App.

Applied

lewri commented on CorsixTH/Lua/dialogs/adviser.lua:
```diff
@@ -58,6 +61,52 @@ function UIAdviser:UIAdviser(ui)
+--! Adds a message to the adviser's message history.
+--! If the history exceeds a certain limit, the oldest messages are removed.
+function UIAdviser:addMessageToHistory(speech)
+  local isMessageInHistory, messageIndexInHistory = self:isMessageInHistory(speech)
```
Camel case doesn't really belong here (line 76). But also why run a boolean when you're already trying to find a value. You can use the existence of a value here for your checks. Also while index can't be 0 in this case, it's weird not to account for it.

```lua
function UIAdviser:getIndexInMessageHistory(message)
  for index, msg in ipairs(self.message_history) do
    if msg == message then
      return index
    end
  end
  return -1
end
```

```lua
local message_history_index = self:getIndexInMessageHistory(speech)

if message_history_index > 0 then
...
```

ARGAMX replied:
Applied

lewri commented on CorsixTH/Lua/dialogs/adviser.lua:
```diff
@@ -233,6 +283,8 @@ function UIAdviser:onMouseDown(button, x, y)
     end
   elseif button == "right" then
     self:hide()
+    --Dump dismissed messages to history
```
Missing a space

ARGAMX replied:
Applied

lewri commented on CorsixTH/Lua/window.lua:
```diff
@@ -311,7 +312,7 @@ function Panel:drawLabel(canvas, x, y, limit)
   if not multi_line then
     text = {text}
-    wrapped = false
+    wrapped = false or self.force_text_wrap
```
Just write `wrapped = self.force_text_wrap`. A default false value is already false.

ARGAMX replied:
Applied

ARGAMX committed:
Merge branch 'master' into pr/3016

ARGAMX commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,187 @@
+
+function UIAdviserHistory:close()
+  Window.close(self)
+end
```
```suggestion
```

ARGAMX commented on CorsixTH/Lua/dialogs/adviser.lua:
```diff
@@ -233,6 +283,8 @@ function UIAdviser:onMouseDown(button, x, y)
     end
   elseif button == "right" then
     self:hide()
+    --Dump dismissed messages to history
```
```suggestion
    -- Dump dismissed messages to history
```

ARGAMX commented on CorsixTH/Lua/dialogs/adviser.lua:
```diff
@@ -58,6 +61,52 @@ function UIAdviser:UIAdviser(ui)
+  if isMessageInHistory then
+    -- If the message is already in history, remove it first
+    table.remove(self.message_history, messageIndexInHistory)
+  end
```
```suggestion
function UIAdviser:messageIndexInHistory(message)
  for index, msg in ipairs(self.message_history) do
    if msg == message then
      return index
    end
  end
  return nil
end

--! Adds a message to the adviser's message history.
--! If the history exceeds a certain limit, the oldest messages are removed.
function UIAdviser:addMessageToHistory(speech)
  local message_index_in_history = self:messageIndexInHistory(speech)

  if message_index_in_history then
    -- If the message is already in history, remove it first
    table.remove(self.message_history, message_index_in_history)
  end
```

ARGAMX commented on CorsixTH/Lua/window.lua:
```diff
@@ -311,7 +312,7 @@ function Panel:drawLabel(canvas, x, y, limit)
   if not multi_line then
     text = {text}
-    wrapped = false
+    wrapped = false or self.force_text_wrap
```
```suggestion
    wrapped = self.force_text_wrap
```

ARGAMX committed:
Apply lewri and tobylane suggestions from code review

Co-authored-by: Artem Argus <ARGAMX@users.noreply.github.com>

ARGAMX committed:
Apply suggestions from code review

Co-authored-by: Artem Argus <ARGAMX@users.noreply.github.com>

ARGAMX commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,184 @@
+        :makeScrollbar(col_bg, scrollbarMovedCallback, 1, 1, 10, 1)
+    -- Add close button
+    self:addPanel(337, x,  8):makeButton(0, 0, 24, 24, 338, self.close)
+        :setTooltip(_S.tooltip.adviser_history.close)
```
```suggestion
    y = self.window_margin + self.header_height + 10

    -- Add scrollbar
    self.scrollbar = self:addColourPanel(x, y, 24, self.row_height * rows - 10,
        col_shadow.red, col_shadow.green, col_shadow.blue)
        :makeScrollbar(col_bg, scrollbarMovedCallback, 1, 1, 10, 1)

    -- Add close button
    self:addPanel(337, x, 15):makeButton(0, 0, 24, 24, 338, self.close)
        :setTooltip(_S.tooltip.adviser_history.close)
```

ARGAMX committed:
Apply suggestions from code review

Co-authored-by: Toby <mrtobylane@gmail.com>
Co-authored-by: Artem Argus <ARGAMX@users.noreply.github.com>

ARGAMX committed:
SAVEGAME_VERSION bump

ARGAMX commented on CorsixTH/Lua/dialogs/adviser.lua:
```diff
@@ -41,6 +41,9 @@ function UIAdviser:UIAdviser(ui)
   self.queued_messages = {}        -- There might be many messages in a row
   self.timer = nil                 -- Timer which hides adviser at the end
 
+  self.message_history = {}        -- Store past messages to be able to view them later
+  self.message_history_limit = 20  -- Limit the number of messages stored
```
Applied

TheCycoONE commented on CorsixTH/Lua/dialogs/resizables/adviser_history.lua:
```diff
@@ -0,0 +1,185 @@
+---@type UIAdviserHistory
+local UIAdviserHistory = _G["UIAdviserHistory"]
+
+local col_bg = { red = 60, green = 174, blue = 201, }
```
Is it possible to use the new colour constants here?

lewri replied:
I don't think they have overlap with the standards, unless you want this window purple. Though adding them as their own definitions in utility is a fine option too.

ARGAMX replied:
Colors moved to utility. 👌

TheCycoONE commented on CorsixTH/Lua/window.lua:
```diff
@@ -311,7 +312,7 @@ function Panel:drawLabel(canvas, x, y, limit)
   if not multi_line then
     text = {text}
-    wrapped = false
+    wrapped = self.force_text_wrap
```
What's the git blame on this one anyway. Seems like there's been a messy history with the wrapping (auto_clip?)

It's strange that passing in the text as a table is the only way to get wrapping.

ARGAMX replied:
> What's the git blame on this one anyway. Seems like there's been a messy history with the wrapping (auto_clip?)
> 
> It's strange that passing in the text as a table is the only way to get wrapping.

Indeed, it is strange.
Git blame points to 92305d15e345f291dfbf07988d01d0559a20c46b.
This logic has not changed since 2013.
What should we do with this?

TheCycoONE replied:
I don't love the force wrap name. I'd like a setWrapped true/false and fall back to the table check if it is nil.

Separately in another PR I'd like everywhere that passes a table to add setWrapped() and I would like auto_clip removed.

ARGAMX replied:
> I don't love the force wrap name. I'd like a setWrapped true/false and fall back to the table check if it is nil.
> 
> Separately in another PR I'd like everywhere that passes a table to add setWrapped() and I would like auto_clip removed.

Okay.
I made it so that if `setWrapped` isn't set, it simply falls back to the behavior that was before this PR.
Also removed `force` from the name.

ARGAMX committed:
Adviser Panel colors moved to utility

ARGAMX committed:
Merge branch 'master' into pr/3016

ARGAMX committed:
Update adviser.lua

ARGAMX committed:
fallback for `wrapped` if `self.wrap_text` not set

lewri commented on CorsixTH/Lua/window.lua:
```diff
@@ -304,17 +305,17 @@ end
     text = {text}
-    wrapped = false
     center_y = true
+    wrapped = (self.wrap_text ~= nil) and self.wrap_text or false
```
Does this break wrapping on any other windows?  I'm guessing nothing currently uses this

ARGAMX replied:
> Does this break wrapping on any other windows? I'm guessing nothing currently uses this

On any other windows it still stay `false` as it was `false`.

ARGAMX committed:
Merge branch 'master' into pr/3016

ARGAMX committed:
Update adviser.lua
